agent/structs/config_entry_gateways.go

```diff
@@ -687,3 +690,300 @@ func (g *GatewayService) Clone() *GatewayService {
+	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
+	RaftIndex
+}
+
+func (e *APIGatewayConfigEntry) GetKind() string {
+	return APIGateway
+}
+
+func (e *APIGatewayConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+	return e.Name
+}
+
+func (e *APIGatewayConfigEntry) GetMeta() map[string]string {
+	if e == nil {
+		return nil
+	}
+	return e.Meta
+}
+
+func (e *APIGatewayConfigEntry) Normalize() error {
+	for i, listener := range e.Listeners {
+		protocol := strings.ToLower(string(listener.Protocol))
+		listener.Protocol = APIGatewayListenerProtocol(protocol)
+		e.Listeners[i] = listener
+	}
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) Validate() error {
+	if err := e.validateListenerNames(); err != nil {
+		return err
+	}
+	if err := e.validateMergedListeners(); err != nil {
+		return err
+	}
+
+	return e.validateListeners()
+}
+
+func (e *APIGatewayConfigEntry) validateListenerNames() error {
+	listeners := make(map[string]struct{})
+	for _, listener := range e.Listeners {
+		if _, found := listeners[listener.Name]; found {
+			return fmt.Errorf("found multiple listeners with the name %q", listener.Name)
+		}
+		listeners[listener.Name] = struct{}{}
+	}
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) validateMergedListeners() error {
+	listeners := make(map[int]APIGatewayListener)
+	for _, listener := range e.Listeners {
+		merged, found := listeners[listener.Port]
+		if found && (merged.Hostname != listener.Hostname || merged.Protocol != listener.Protocol) {
+			return fmt.Errorf("listeners %q and %q cannot be merged", merged.Name, listener.Name)
+		}
+		listeners[listener.Port] = listener
+	}
+	return nil
+}
+
+func (e *APIGatewayConfigEntry) validateListeners() error {
+	validProtocols := map[APIGatewayListenerProtocol]bool{
+		ListenerProtocolHTTP: true,
+		ListenerProtocolTCP:  true,
+	}
+	allowedCertificateKinds := map[string]bool{
+		InlineCertificate: true,
+	}
+
+	for _, listener := range e.Listeners {
+		if !validProtocols[listener.Protocol] {
+			return fmt.Errorf("unsupported listener protocol %q, must be one of 'tcp', or 'http'", listener.Protocol)
+		}
+		if listener.Protocol == ListenerProtocolTCP && listener.Hostname != "" {
+			// TODO: once we have SNI matching we should be able to implement this
+			return fmt.Errorf("hostname specification is not supported when using TCP")
+		}
+		if listener.Port <= 0 || listener.Port > 65535 {
+			return fmt.Errorf("listener port %d not in the range 1-65535", listener.Port)
+		}
+		if strings.ContainsRune(strings.TrimPrefix(listener.Hostname, wildcardPrefix), '*') {
+			return fmt.Errorf("host %q is not valid, a wildcard specifier is only allowed as the left-most label", listener.Hostname)
+		}
+		for _, certificate := range listener.TLS.Certificates {
+			if !allowedCertificateKinds[certificate.Kind] {
+				return fmt.Errorf("unsupported certificate kind: %q, must be 'inline-certificate'", certificate.Kind)
+			}
+			if certificate.Name == "" {
```

Comment:
In enterprise, will we need to validate the namespace/partition too or will we restrict to referencing certificates in the same ns/partition. 

Having give this a whole 5 seconds of consideration it seems fine to restrict to the same ns/partition as the API GW itself but maybe you know of a reason not to.

I guess for this bit of code we won't need to validate it regardless and that would be handled by the controller when binding a cert to the gw right?

Reply:
I do believe that we have to support cross-namespace/partition certificate references since K8s Gateway API allows them today, and an Enterprise user with namespace mirroring turned on would want to be able to use this.

The pattern that the K8s Gateway API uses repeatedly is: if the namespace is unset, then assume the referenced object lives in the same namespace as the referencing object. Thoughts on using that same approach here @mkeeler ? The implication for this particular bit of code is that we wouldn't need to validate that the namespace is set, we would resolve everything at binding time as you mentioned.

Reply:
Defaulting to the same ns as the existing object is how our other config entries work (splitters, routers, resolvers) so it seems good to continue that pattern.

In general, I think cross-ns references are probably okay. There are limitations with consul ACLs that could make that a little tricky if the ACL token used for the API GW itself is not in the default ns but we could go over that more outside of github comments.

I still think it could make sense to not allow cross-partition references to certificates. In general in order to do that we would need some new ACL enforcement mechanism. The only thing we allow cross-partition access is exported services. In that case an operator has explicitly exported a service to another partition. Unless we do something similar or bake in some support in the ACL rules themselves, cross-partition references should be prevented.

Reply:
Added this conversation to "Matt's Thoughts About the Future": https://docs.google.com/document/d/1SmsakmGnrmd3uI3mthefHQkDrz6GvzMrQvaHoiKu15Y/edit#

I want to make sure we capture these and implement them while moving forward with this existing work.

Reply:
Thanks guys!